ProjFS.Mac/PrjFSKextTests/VnodeCacheTests.mm

```diff
@@ -0,0 +1,745 @@
+        &self->dummyPerfTracer,
+        PrjFSPerfCounter_VnodeOp_Vnode_Cache_Hit,
+        PrjFSPerfCounter_VnodeOp_Vnode_Cache_Miss,
+        PrjFSPerfCounter_VnodeOp_FindRoot,
+        PrjFSPerfCounter_VnodeOp_FindRoot_Iteration,
+        self->testVnodeFile1.get(),
+        self->dummyVFSContext));
+    
+    uintptr_t vnodeIndex = ComputeVnodeHashIndex(self->testVnodeFile1.get());
+    XCTAssertTrue(self->testVnodeFile1.get() == cacheWrapper[vnodeIndex].vnode);
+    XCTAssertTrue(self->testVnodeFile1->GetVid() == cacheWrapper[vnodeIndex].vid);
+    XCTAssertTrue(repoRootHandle == cacheWrapper[vnodeIndex].virtualizationRoot);
+    
+    XCTAssertTrue(repoRootHandle == VnodeCache_FindRootForVnode(
+        &self->dummyPerfTracer,
+        PrjFSPerfCounter_VnodeOp_Vnode_Cache_Hit,
+        PrjFSPerfCounter_VnodeOp_Vnode_Cache_Miss,
+        PrjFSPerfCounter_VnodeOp_FindRoot,
+        PrjFSPerfCounter_VnodeOp_FindRoot_Iteration,
+        self->testVnodeFile1.get(),
+        self->dummyVFSContext));
+    
+    // Sanity check: We don't expect any of the mock functions to have been called
+    XCTAssertFalse(MockCalls::DidCallAnyFunctions());
+}
+
+- (void)testVnodeCache_RefreshRootForVnode {
+    VirtualizationRootHandle repoRootHandle = InsertVirtualizationRoot_Locked(
+        nullptr /* no client */,
+        0,
+        self->repoRootVnode.get(),
+        self->repoRootVnode->GetVid(),
+        FsidInode{ self->repoRootVnode->GetMountPoint()->GetFsid(), self->repoRootVnode->GetInode() },
+        self->repoPath.c_str());
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(repoRootHandle));
+
+    VirtualizationRootHandle foundRoot = VirtualizationRoot_FindForVnode(
+        &self->dummyPerfTracer,
+        PrjFSPerfCounter_VnodeOp_FindRoot,
+        PrjFSPerfCounter_VnodeOp_FindRoot_Iteration,
+        self->testVnodeFile1.get(),
+        self->dummyVFSContext);
+    XCTAssertEqual(foundRoot, repoRootHandle);
+
+    // We don't care which mocks were called during the initialization code (above)
+    MockCalls::Clear();
+
+    // Initialize the cache
+    VnodeCacheEntriesWrapper cacheWrapper(/* capacity*/ 100, /* fillCache*/ false);
+    
+    // Insert testFileVnode with TestRootHandle as its root
+    uintptr_t indexFromHash = ComputeVnodeHashIndex(self->testVnodeFile1.get());
+    uint32_t testVnodeVid = self->testVnodeFile1->GetVid();
+    XCTAssertTrue(
+        TryInsertOrUpdateEntry_ExclusiveLocked(
+            self->testVnodeFile1.get(),
+            indexFromHash,
+            testVnodeVid,
+            false, // forceRefreshEntry
+            TestRootHandle));
+    XCTAssertTrue(testVnodeVid == cacheWrapper[indexFromHash].vid);
+    XCTAssertTrue(TestRootHandle == cacheWrapper[indexFromHash].virtualizationRoot);
+    
+    // VnodeCache_RefreshRootForVnode should
+    // force a lookup of the new root and set it in the cache
+    VirtualizationRootHandle rootHandle = VnodeCache_RefreshRootForVnode(
```

Comment:
Does this validate that the vnode was actually refreshed in the cache?  Seems like that may be hard to do.

Reply:
That's what it should be validating, but I see that we need another check to validate that `rootHandle` is not the same as `TestRootHandle`.